platforms/core-runtime/launcher/src/main/java/org/gradle/launcher/daemon/protocol/DaemonMessageSerializer.java

```diff
@@ -194,7 +200,40 @@ private static class BuildEventSerializer implements Serializer<BuildEvent> {
 
         @Override
         public void write(Encoder encoder, BuildEvent buildEvent) throws Exception {
-            payloadSerializer.write(encoder, buildEvent.getPayload());
+            if (buildEvent.getPayload() instanceof DefaultTaskStartedProgressEvent) {
```

Comment:
I think it would be better to do the truncation where the objects are created, rather than making a copy here. This avoids baking knowledge of how the descriptor will be used into this serializer, which really should just send whatever data it is given, and also avoids generating a whole bunch of extra garbage in creating the copies.

To do this, I would probably change `OperationDependenciesResolver` to do the truncation, so that we also solve the same problem for the events sent for artifact transforms (there can be many more transforms than tasks).

Reply:
Well, my reasoning is, as I do not know all the internal details of Gradle, that this full information might actually be used somewhere else.
I am only sure that sending all this to the client is 1) inefficient and 2) unnecessary as there it is restored from cached descriptors. That's why I considered it to be a serialization problem.

Thanks for the pointer, I will take a look! If my considerations above prove to be invalid I'll happily move the fix, I don't really like the current 'hack' fix as well.

Reply:
Agreed, it is indeed much more natural to change in the suggested area.
Though I think the most natural is to create two versions of descriptors where we create them and cache the one without dependencies. Please take a look.